File: demo-site/src/lib/biometrics/keystroke.ts
/**
 * TypeScript port of the legacy (no-pretrained-model) fallback path in
 * `behavioral_biometrics_nn/modules/keystroke.py`'s `KeystrokeModule`: a
 * pure online z-score baseline over dwell/flight times, learned during
 * warm-up. Chosen deliberately over the module's *primary* path (a second
 * pretrained Siamese encoder + per-session gallery, mirroring the mouse
 * module) to avoid exporting/bundling a second neural net for this
 * serverless MVP - this fallback is a real, tested code path in the local
 * demo (used whenever `artifacts/keystroke/` hasn't been trained), not an
 * ad hoc simplification.
 */

import type { ModuleOutcome } from "./tabNavigation";

export type KeystrokeState = {
  dwellBaseline: number[];
  flightBaseline: number[];
  nKeys: number;
};

export function initKeystrokeState(): KeystrokeState {
  return { dwellBaseline: [], flightBaseline: [], nKeys: 0 };
}

const MIN_KEYS_PER_WINDOW = 4;
const KEYS_FOR_FULL_CONFIDENCE = 6; // ~72 WPM; reaching this in a 1 s window gives full authority
const WARMUP_KEYS = 40;
const CORRECTION_KEYS = new Set(["Backspace", "Delete"]);

function zscore(value: number, baseline: number[]): number {
  if (baseline.length < 3) return 0;
  const m = baseline.reduce((a, b) => a + b, 0) / baseline.length;
  const variance = baseline.reduce((a, b) => a + (b - m) ** 2, 0) / baseline.length;
  const std = Math.sqrt(variance);
  return std < 1e-9 ? 0 : (value - m) / std;
}

export function updateKeystroke(
  state: KeystrokeState,
  keys: { key: string; dwell: number; flight: number }[],
): ModuleOutcome {
  const dwells = keys.map((k) => Number(k.dwell)).filter(Number.isFinite);
  const flights = keys.map((k) => Number(k.flight)).filter(Number.isFinite);

  if (dwells.length < MIN_KEYS_PER_WINDOW) {
    return {
      risk: 0,
      confidence: 0,
      status: "inactive",
      detail: { reason: `only ${dwells.length} keystrokes (need ${MIN_KEYS_PER_WINDOW})` },
    };
  }

  const corrections = keys.filter((k) => CORRECTION_KEYS.has(String(k.key))).length;
  const confidence = Math.min(1.0, dwells.length / KEYS_FOR_FULL_CONFIDENCE);

  state.nKeys += dwells.length;
  const errorRatio = corrections / Math.max(keys.length, 1);

  if (state.nKeys < WARMUP_KEYS) {
    state.dwellBaseline.push(...dwells);
    state.flightBaseline.push(...flights);
    return { risk: 0, confidence: 0, status: "warming", detail: { keys_seen: state.nKeys, need: WARMUP_KEYS } };
  }

  const dwellMean = dwells.reduce((a, b) => a + b, 0) / dwells.length;
  const flightMean = flights.length ? flights.reduce((a, b) => a + b, 0) / flights.length : 0;
  const dwellZ = Math.abs(zscore(dwellMean, state.dwellBaseline));
  const flightZ = flights.length ? Math.abs(zscore(flightMean, state.flightBaseline)) : 0;

  const risk = Math.min(
    1.0,
    0.45 * Math.min(dwellZ / 3.0, 1.0) + 0.45 * Math.min(flightZ / 3.0, 1.0) + 0.1 * Math.min(errorRatio * 2.0, 1.0),
  );

  return {
    risk,
    confidence,
    status: "active",
    detail: {
      dwell_zscore: dwellZ,
      flight_zscore: flightZ,
      error_rate_ratio: errorRatio,
      keys_this_window: dwells.length,
    },
  };
}

File: behavioral_biometrics_nn/modules/keystroke.py
"""Extension B: keystroke dynamics (flight / dwell times).

The Balabit training data is mouse-only, so this module cannot be pre-trained
on it. It learns the owner's typing rhythm online during warm-up and scores
later windows as a z-score deviation from that baseline.
"""

from __future__ import annotations

import math
from statistics import fmean, pstdev

from .base import STATUS_ACTIVE, STATUS_WARMING, ModuleResult, RiskModule

WARMUP_KEYS = 40
MIN_KEYS_PER_WINDOW = 4
CORRECTION_KEYS = frozenset({"Backspace", "Delete"})


class KeystrokeModule(RiskModule):
    plugin_id = "keystroke_v1"
    display_name = "Keystroke dynamics"
    default_weight = 0.5

    def __init__(self, warmup_keys: int = WARMUP_KEYS):
        self.warmup_keys = warmup_keys
        self.reset()

    def reset(self) -> None:
        self.dwell_baseline: list[float] = []
        self.flight_baseline: list[float] = []
        self.n_keys = 0

    def update(self, telemetry: dict) -> ModuleResult:
        keys = telemetry.get("keys", [])
        dwells = [float(k["dwell"]) for k in keys if _is_number(k.get("dwell"))]
        flights = [float(k["flight"]) for k in keys if _is_number(k.get("flight"))]

        if len(dwells) < MIN_KEYS_PER_WINDOW:
            return self.inactive(f"only {len(dwells)} keystrokes (need {MIN_KEYS_PER_WINDOW})")

        self.n_keys += len(dwells)
        corrections = sum(1 for k in keys if str(k.get("key", "")) in CORRECTION_KEYS)

        if self.n_keys < self.warmup_keys:
            self.dwell_baseline.extend(dwells)
            self.flight_baseline.extend(flights)
            return ModuleResult(
                self.plugin_id, 0.0, 0.0, STATUS_WARMING,
                {"keys_seen": self.n_keys, "need": self.warmup_keys},
            )

        dwell_z = abs(_zscore(fmean(dwells), self.dwell_baseline))
        flight_z = abs(_zscore(fmean(flights), self.flight_baseline)) if flights else 0.0
        error_ratio = corrections / max(len(keys), 1)

        risk = min(
            1.0,
            0.45 * min(dwell_z / 3.0, 1.0)
            + 0.45 * min(flight_z / 3.0, 1.0)
            + 0.10 * min(error_ratio * 2.0, 1.0),
        )
        confidence = min(1.0, len(dwells) / (2 * MIN_KEYS_PER_WINDOW))

        return ModuleResult(
            self.plugin_id, risk, confidence, STATUS_ACTIVE,
            {
                "dwell_zscore": round(dwell_z, 3),
                "flight_zscore": round(flight_z, 3),
                "error_rate_ratio": round(error_ratio, 3),
                "keys_this_window": len(dwells),
            },
        )


def _zscore(value: float, baseline: list[float]) -> float:
    if len(baseline) < 3:
        return 0.0
    std = pstdev(baseline)
    return 0.0 if std < 1e-9 else (value - fmean(baseline)) / std


def _is_number(value) -> bool:
    try:
        return math.isfinite(float(value))
    except (TypeError, ValueError):
        return False
